docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,555 @@
+CRD spec.
+
+```go
+// model.go
+// GetCRDs returns a slice of `CRD` structs that describe the
+// top-level resources discovered by the code generator for an AWS service API
+func (m *Model) GetCRDs() ([]*CRD, error) {
+	if m.crds != nil {
+		return m.crds, nil
+	}
+	crds := []*CRD{}
+
+	opMap := m.SDKAPI.GetOperationMap(m.cfg)
+
+	createOps := (*opMap)[OpTypeCreate]
+	readOneOps := (*opMap)[OpTypeGet]
+	readManyOps := (*opMap)[OpTypeList]
+	updateOps := (*opMap)[OpTypeUpdate]
+	deleteOps := (*opMap)[OpTypeDelete]
+	getAttributesOps := (*opMap)[OpTypeGetAttributes]
+	setAttributesOps := (*opMap)[OpTypeSetAttributes]
+
+	for crdName, createOp := range createOps {
+		if m.cfg.IsIgnoredResource(crdName) {
+			continue
+		}
+		crdNames := names.New(crdName)
+...
+			crd.AddSpecField(memberNames, memberShapeRef)
+
++			fConfig := m.cfg.ResourceFields(crdName)[fieldName]
++			if fConfig != nil && fConfig.Reference != nil {
++				referenceFieldNames := names.New(fieldName + "Ref")
++				rf := NewReferenceField(crd, referenceFieldNames, memberShapeRef)
++				crd.SpecFields[referenceFieldNames.Original] = rf
++				crd.Fields[referenceFieldNames.Camel] = rf
++			}
+		}
+
+		// Now any additional Spec fields that are required from other API
+		// operations.
+		for targetFieldName, fieldConfig := range m.cfg.ResourceFields(crdName) {
+			if fieldConfig.IsReadOnly {
+				// It's a Status field...
+				continue
+			}
+
+...
+	m.crds = crds
+	return crds, nil
+}
+```
+
+```go
+//field.go
+func NewReferenceField(
+	crd *CRD,
+	fieldNames names.Names,
+	shapeRef *awssdkmodel.ShapeRef,
+) *Field {
+	gt := "*ackv1alpha1.AWSResourceReferenceWrapper"
+	gtp := "*github.com/aws-controllers-k8s/runtime/apis/core/v1alpha1.AWSResourceReferenceWrapper"
+	gte := ""
+	if shapeRef.Shape.Type == "list" {
+		gt = "*ackv1alpha1.AWSResourcesReferenceWrapper"
+		gtp = "*github.com/aws-controllers-k8s/runtime/apis/core/v1alpha1.AWSResourcesReferenceWrapper"
+		gte = "*ackv1alpha1.AWSResourcesReferenceWrapper"
+	}
+	return &Field{
+		CRD:               crd,
+		Names:             fieldNames,
+		Path:              fieldNames.Original,
+		ShapeRef:          nil,
+		GoType:            gt,
+		GoTypeElem:        gte,
+		GoTypeWithPkgName: gtp,
+		FieldConfig:       nil,
+	}
+}
+```
+
+#### Runtime Changes
+
+```go
+// aws_resource_manager.go
+
+type AWSResourceManager interface {
+        ...
++       // ResolveReferences finds if there are any AWSResourceReference fields
++       // present in the AWSResource passed in the parameters and attempts to resolve
++       // those reference fields into target field.
++       // It returns an AWSResource with resolved references, boolean representing
++       // whether any reference fields were present and an error if the passed
++       // AWSResource's reference fields cannot be resolved
++       ResolveReferences(context.Context, client.Reader, AWSResource) (AWSResource, bool, error)
```

Comment:
nit: I think the `bool` return value is unnecessary here because returning a nil reference (changing the first return value to a pointer) would suffice.

Reply:
Good suggestion, but we will be introducing pointer to an interface. We have not used pointer to interface before in runtime and I am not sure If it's worth it to remove a return type. 

I'll see what Jay or Amine think about it.

Reply:
+1 to removing bool, interface will be simpler and more consistent with existing runtime methods like `updateResource` and `reconcile`

Reply:
What about making a new `AWSResourceResolver` interface, inheriting from `AWSResource` and `ConditionManager` and just returning `error`?

```go
type AWSResourceResolver interface {
       AWSResource
       ConditionManager
       // ResolveReferences finds if there are any AWSResourceReference fields
       // present in the AWSResource and returns an error if the
       // reference fields cannot be resolved
       ResolveReferences(context.Context, client.Reader) error
}
```

There's no reason to have the copies of desired/latest, IMHO. Just have the `AWSResourceResolver.ResolveReferences` mutate the resource as needed to resolve the referenced resource identifiers and add the appropriate `Condition` to the resource itself (which it can do because it inherits from `ConditionManager`.

Within the reconciler's `Sync` method, you would simplify the code do just this:

```go
	if err := desired.ResolveReferences(ctx, r.apiReader); err != nil {
		return desired, err
	}
```

Reply:
Done. 
I changed the return type of `ResolveReferences` method to `(AWSResource, error)` and updated the conditions within that method.

@jaypipes , your suggestion was very good but when I started to implement it, I found that I will need to change type of `desired` to new interface to be able to call `desired.ResolveReferences` method.
**More importantly**, if we change the desired(by resolving references), the `PatchStatus` method will not see the new conditions as diff between `desired` & `latest` and these condition will not be persisted in K8s cluster.

Since we need a new `AWSResource` after resolving references, I sticked to existing `AWSResourceManager.ResolveReferences()` method and was able to cleanly update conditions within the method. 

PTAL.